### PR TITLE
Merging dev-gateway into dev

### PR DESCRIPTION
**Pull request by itsVale**
# Pull Request Template

Please try and complete the below as best as possible. Some of the fields may
not be necessary so feel free to add or edit as you see fit.

## Checklist

_Confirm you have completed the following actions prior to submitting this PR._

- [ ] There is an existing issue report for this PR.
- [ ] I have forked this project.
- [x] I have created a feature branch.
- [x] My changes have been committed.
- [x] My changes were linted using pylava.
- [x] I have pushed my changes to the branch.

## Title

_Give your PR a short title summarising the patch, bug fix or feature._
Merging the `dev-gateway` branch into `dev`

## Description

_Ensure the PR description clearly describes the problem and solution and provide as much relevant information as possible._
After doing some minor refactoring, I also fixed all the linting issues and I think we're ready now to get rid of the `dev-gateway` branch.

## Issue Resolution

_Tell us which issue this PR fixes._

This Pull Request Fixes #

## Proposed Changes

_List your proposed changes below._

- This PR fixes a bug relating to...
- Adding a patch that enhances the project.
- Various bug fixes.

Way too many to list. Check out the Commits haha.

## New or Changed Features

_Does this PR provide new or changed features or enhancements? If so, what is included in this PR?_

Adding a Client that is able to interact with the REST API as well as with the Gateway, some basic model classes and support for a few Gateway events. Also a bit of internal caching.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
     long_description_content_type='text/markdown',
     url='https://github.com/itsVale/Shitcord',
     license='GNU General Public License v3 (GPLv3)',
-    packages=['shitcord', 'shitcord.http'],
+    packages=['shitcord', 'shitcord.http', 'shitcord.gateway', 'shitcord.models', 'shitcord.events', 'shitcord.utils'],
```
Comment (romangraef) [inline setup.py]: Maybe use find_packages for that. not exactly sure.

File: shitcord/__init__.py [hunk 2]
```diff
@@ -12,19 +12,27 @@
...
 __url__ = 'https://github.com/itsVale/Shitcord'
 
-from .http.api import API
-
-import logging
-import sys
-from collections import namedtuple
-
 Version = namedtuple('VersionInfo', 'major minor micro releaselevel')
 version_info = Version(major=0, minor=0, micro=1, releaselevel='beta')
```
Comment (romangraef) [inline shitcord/__init__.py]: why is micro still =1 not =2

File: shitcord/gateway/serialization.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import functools
+import json
+import sys
+
+import shitcord
+from .opcodes import Opcodes
+from ..utils.jsonenum import EnumEncoder
```
Comment (romangraef) [inline shitcord/gateway/serialization.py]: maybe we should resolve those to from shitcord. etc imports

File: shitcord/http/http.py [hunk 4]
```diff
@@ -2,8 +2,8 @@
 
 import logging
 import shitcord
-from shitcord.http import rate_limit
-from shitcord.http.errors import ShitRequestFailedError
+from . import rate_limit
+from .errors import ShitRequestFailedError
```
Comment (romangraef) [inline shitcord/http/http.py]: Are we intentionally using from . imports or should we stay with from shitcode.http imports

File: shitcord/models/channel.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+from shitcord.models.core import Model
+
+
+class Channel(Model):
+    def __init__(self, data):
+        super().__init__(data)
+
+        self.guild_id = data['guild_id']
```
Comment (romangraef) [inline shitcord/models/channel.py]: guild id is optional

File: shitcord/models/channel.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        self.name = data['name']
+        self.permissions_overwrites = data['permission_overwrites']
+        self.topic = data['topic']
+        self.parent_id = int(data['parent_id'])
```
Comment (romangraef) [inline shitcord/models/channel.py]: parent id is optional

File: shitcord/models/channel.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        self.nsfw = bool(data['nsfw'])
+        self.position = int(data['position'])
```
Comment (romangraef) [inline shitcord/models/channel.py]: position is optional

File: shitcord/models/invite.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+from shitcord.models.core import Model
+from shitcord.models.guild import Guild
+
+
+class Invite(Model):
+    def __init__(self, data):
+        super().__init__(data)
+
+        self.code = data['code']
+        self.guild = Guild(data['guild'])
```
Comment (romangraef) [inline shitcord/models/invite.py]: guild is a partial guild object, and not necessarily included

File: shitcord/models/webhook.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+from shitcord.models.core import Model
+from shitcord.models import User
+
+
+class Webhook(Model):
+    def __init__(self, data):
+        super().__init__(data)
+
+        self.id = data['id']
+        self.guild_id = data['guild_id']
```
Comment (romangraef) [inline shitcord/models/webhook.py]: guild id is optional

File: shitcord/models/webhook.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        self.channel_id = data['channel_id']
+        self.name = data['name']
+        self.avatar = data['avatar']
+        self.token = data['token']
+        try:
+            self.user = User(data['user'])
+        except KeyError:
+            self.user = None
```
Comment (romangraef) [inline shitcord/models/webhook.py]: maybe try using 
```py
self.user = User(data['user']) if 'user' in data else None
```
exceptions are (usually) for exceptional situations

Comment (romangraef) [inline shitcord/models/webhook.py]: maybe try using 
```py
self.user = User(data['user']) if 'user' in data else None
```
exceptions are (usually) for exceptional situations

File: shitcord/models/channel.py (hunk 5 above)
Comment (itsVale) [reply shitcord/models/channel.py]: I didn't focus too much on the models as they are just placeholders for now. They have no real usage. I started to recreate them a few hours before and will continue to do that the next few days. I'm aware of all those issues with the optional values.

File: shitcord/__init__.py (hunk 2 above)
Comment (itsVale) [reply shitcord/__init__.py]: Fixed that. Thanks.

File: setup.py (hunk 1 above)
Comment (itsVale) [reply setup.py]: Will try find_packages over the days, I already had that thought. But for now, this is good enough.

File: shitcord/http/http.py (hunk 4 above)
Comment (itsVale) [reply shitcord/http/http.py]: Let's stick with that . syntax for imports.

**Pull request merged**